File: crawler/src/engine/deduper.ts
import config from "../config";

const SIMHASH_BITS = 64;

function tokenize(text: string): string[] {
  const cleaned = text.replace(/[\s，。！？、；：""''（）《》【】]/g, " ");
  const tokens: string[] = [];
  const chars = cleaned.replace(/\s/g, "");

  for (let i = 0; i < chars.length - 1; i++) {
    tokens.push(chars.slice(i, i + 2));
  }

  if (tokens.length === 0 && chars.length > 0) {
    tokens.push(chars);
  }

  return tokens;
}

function stringHash(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const chr = str.charCodeAt(i);
    hash = (hash << 5) - hash + chr;
    hash |= 0;
  }
  return Math.abs(hash);
}

function intToBinaryString(num: number, bits: number): string {
  return num.toString(2).padStart(bits, "0").slice(-bits);
}

export function computeSimHash(text: string): string {
  if (!text) return "";

  const tokens = tokenize(text);
  const v: number[] = new Array(SIMHASH_BITS).fill(0);

  for (const token of tokens) {
    const hash = stringHash(token);
    const binary = intToBinaryString(hash, SIMHASH_BITS);

    for (let i = 0; i < SIMHASH_BITS; i++) {
      if (binary[i] === "1") {
        v[i] += 1;
      } else {
        v[i] -= 1;
      }
    }
  }

  let fingerprint = "";
  for (let i = 0; i < SIMHASH_BITS; i++) {
    fingerprint += v[i] >= 0 ? "1" : "0";
  }

  return fingerprint;
}

export function computeFingerprint(text: string): string {
  if (!text) return "";
  return computeSimHash(text);
}

export function hammingDistance(hash1: string, hash2: string): number {
  if (hash1.length !== hash2.length) {
    throw new Error("Hash lengths must match");
  }
  let distance = 0;
  for (let i = 0; i < hash1.length; i++) {
    if (hash1[i] !== hash2[i]) distance++;
  }
  return distance;
}

export function isDuplicate(
  newHash: string,
  existingHashes: string[],
  threshold: number = config.simhashThreshold
): boolean {
  for (const existing of existingHashes) {
    if (!existing) continue;
    if (hammingDistance(newHash, existing) <= threshold) {
      return true;
    }
  }
  return false;
}

File: crawler-py/crawler/engine/deduper.py
"""去重引擎

对应 TypeScript 版 crawler/src/engine/deduper.ts。
直接复用 server-py 的 app/utils/dedup.py（MD5 + SimHash），
这里做一层同名包装，保持与 TS 版一致的对外接口。
"""
import os
import sys

_SERVER_PY = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "server-py"))
if _SERVER_PY not in sys.path:
    sys.path.insert(0, _SERVER_PY)

from app.utils.dedup import dedup  # noqa: E402

from ..config import config  # noqa: E402

SIMHASH_BITS = dedup["SIMHASH_BITS"]


def compute_sim_hash(text: str) -> str:
    """对应 TS computeSimHash"""
    if not text:
        return ""
    return dedup["computeSimHash"](text)


def compute_fingerprint(text: str) -> str:
    """对应 TS computeFingerprint（SimHash 指纹）"""
    if not text:
        return ""
    return dedup["computeSimHash"](text)


def hamming_distance(hash1: str, hash2: str) -> int:
    """对应 TS hammingDistance（长度不一致时抛异常）"""
    if len(hash1) != len(hash2):
        raise ValueError("Hash lengths must match")
    return sum(1 for a, b in zip(hash1, hash2) if a != b)


def is_duplicate(new_hash: str, existing_hashes: list, threshold: int = None) -> bool:
    """对应 TS isDuplicate"""
    if threshold is None:
        threshold = config.simhash_threshold
    for existing in existing_hashes:
        if not existing:
            continue
        if hamming_distance(new_hash, existing) <= threshold:
            return True
    return False


# 保持 camelCase 别名，方便与 TS 对照
computeSimHash = compute_sim_hash
computeFingerprint = compute_fingerprint
hammingDistance = hamming_distance
isDuplicate = is_duplicate
